### PR TITLE
Add serde support for networks.

### PR DESCRIPTION
**Pull request by craftit**
Add serde support for learnable parameters.
Add an example of training a network, saving it and loading it.

### CODE REVIEW

**Comment by frjnn** [comment]
Will review asap.

File: src/lib.rs [hunk 1]
```diff
@@ -175,8 +175,9 @@ use ndarray_rand::RandomExt;
 pub use variable::{
     Backward, Cat, Convolve, ConvolveWithGroups, Data, Eval, Forward, Gradient, MatMatMul,
     MatMatMulT, Overwrite, Param, Stack, Var, VarDiff, VecMatMul, VecVecMul,
+    Input
 };
-use variable::{Input, InputBackward};
```
Comment (frjnn) [inline src/lib.rs:179]: Input shouldn't be in the public API. I'll fix it before merging on main.

Comment (frjnn) [inline src/lib.rs:179]: `Input` should not be in the public API. I'll fix it before merging on main.

Comment (craftit) [reply src/lib.rs:179]: How are you supposed to load existing arrays? Maybe I am missing something.

File: examples/simple.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+     {
+         let out1 = self.lin1.forward(input).relu();
+         let out2 = self.lin2.forward(out1).relu();
+         let out3 = self.lin3.forward(out2);
+         out3
+    }
+}
+
+/// Load model from a string
+fn load_model() -> MLP {
```
Comment (frjnn) [inline examples/simple.rs:66]: You could probably do it like this, without any addition to the original API: 

```rust
fn load_model() -> MLP {
    let MLP { lin1, lin2, lin3, status } = serde_json::from_str(
             r#"{"lin1":{"weight":{"v":1,"dim":[5,3],"data":[0.31398147,-0.02374097,-0.045672387,-0.57606286,0.5287176,-0.038059983,-0.19196294,0.9338395,-0.34874597,-0.08579302,-0.21880743,0.26289353,0.12593554,-0.19557185,-0.6770759]},"bias":{"v":1,"dim":[5],"data":[0.036578782,-0.3663301,-0.23192844,0.1254652,0.5213851]}},"lin2":{"weight":{"v":1,"dim":[5,5],"data":[0.52091986,0.3500197,-0.06102618,-0.43995684,0.53706765,-0.09257236,-0.3584929,-0.43666622,0.43744308,-0.40631944,0.066774696,0.16129021,-0.25963476,0.26902968,0.1528883,0.12935583,-0.2496377,0.14702061,-0.012540738,-0.34052926,0.45684096,-0.12884608,0.21005273,-0.7786633,-0.08895902]},"bias":{"v":1,"dim":[5],"data":[0.6071196,-0.18910336,-0.2278286,0.044481196,0.10841279]}},"lin3":{"weight":{"v":1,"dim":[1,5],"data":[0.21673596,-0.021770507,-0.00067504647,0.5252394,0.06640336]},"bias":{"v":1,"dim":[1],"data":[0.7723236]}}}"#,
     ).unwrap();

    MLP {
             lin1: status.register(lin1),
             lin2: status.register(lin2),
             lin3: status.register(lin3),
             status,
     }
}
```

File: src/nn/mod.rs [hunk 3]
```diff
@@ -438,6 +441,16 @@ impl ModelStatus {
         component
     }
 
+    /// Registers a component.
```
Comment (frjnn) [inline src/nn/mod.rs:444]: This may be not needed.

**Review by frjnn — COMMENTED** [review]
Looks good to me, would you be willing to add some tests? @craftit

**Comment by craftit** [comment]
I can add some more tests,  the sample works as part of one but I wanted to check the rest of it was ok first.

File: src/lib.rs (hunk 1 above)
Comment (frjnn) [reply src/lib.rs:179]: You can use [neuronika::from_ndarray](https://docs.rs/neuronika/latest/neuronika/fn.from_ndarray.html).

**Comment by frjnn** [comment]
LGMT, should I merge?

**Comment by craftit** [comment]
> LGTM, should I merge?

I've just fixed some warnings I had to make some minor changes to a few things in the core library to get the tests to run, so you may want to review them.

**Pull request merged**